CoreFoundation/RunLoop.subproj/CFRunLoop.c

```diff
@@ -2874,6 +2874,16 @@ static int32_t __CFRunLoopRun(CFRunLoopRef rl, CFRunLoopModeRef rlm, CFTimeInter
 #if TARGET_OS_WIN32 || TARGET_OS_LINUX
             void *msg = 0;
 #endif
+
+#if TARGET_OS_WIN32
+            // Dispatch sets this event to notify us about
+            // pending work on main queue.
+            // We should consume it properly, or CFRunLoop will
+            // immediately poke empty main queue on next run(s)
+            // and exit without any actual work done.
+            ResetEvent(dispatchPort);
```

Comment:
Ugh, this is stupidly subtle.  The corresponding `SetEvent` occurs in libdispatch, not in Foundation.  Because we have just woken up to service libdispatch's main queue, it does need to be reset.  I think that the thing to consider here is if the current dispatch behaviour is correct or not, and perhaps libdispatch should use an auto-reset event instead.